From e0fa993c7309bfff29d3264064d45daadf313424 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:26:00 -0400
Subject: [tools] perf/symbols: Fix a memory leak due to symbol__delete not being used

Message-id: <1408462094-14194-192-git-send-email-jolsa@redhat.com>
Patchwork-id: 88184
O-Subject: [PATCH RHEL7.1 BZ1131394 191/325] perf symbols: Fix a memory leak due to symbol__delete not being used
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit d4f74eb89199dc7bde5579783e9188841e1271e3
Author: Chenggang Qin <chenggang.qcg@taobao.com>
Date: Fri Oct 11 08:27:59 2013 +0800

description
===========
In function symbols__fixup_duplicate(), while duplicated symbols are
found, only the rb_node is removed from the tree. The symbol structures
themself are ignored.  Then, these memory areas are lost.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/symbol.c b/tools/perf/util/symbol.c
index b66c1ee..c0c3696 100644
--- a/tools/perf/util/symbol.c
+++ b/tools/perf/util/symbol.c
@@ -160,10 +160,12 @@ again:
 
   if (choose_best_symbol(curr, next) == SYMBOL_A) {
    rb_erase(&next->rb_node, symbols);
+   symbol__delete(next);
    goto again;
   } else {
    nd = rb_next(&curr->rb_node);
    rb_erase(&curr->rb_node, symbols);
+   symbol__delete(curr);
   }
  }
 }
-- 
1.7.1